stdlib/universe/key_values.go

```diff
@@ -195,44 +206,38 @@ func (t *keyValuesTransformation) Process(id execute.DatasetID, tbl flux.Table)
 	}
 
 	var (
-		nullDistinct   bool
-		boolDistinct   map[bool]bool
-		intDistinct    map[int64]bool
-		uintDistinct   map[uint64]bool
-		floatDistinct  map[float64]bool
-		stringDistinct map[string]bool
-		timeDistinct   map[execute.Time]bool
+		boolDistinct = map[struct {
+			string
+			bool
+		}]bool{}
+		intDistinct = map[struct {
+			string
+			int64
+		}]bool{}
+		uintDistinct = map[struct {
+			string
+			uint64
+		}]bool{}
+		floatDistinct = map[struct {
+			string
+			float64
+		}]bool{}
+		timeDistinct = map[struct {
+			string
+			execute.Time
+		}]bool{}
+		nullDistinct   = false
+		stringDistinct = map[[2]string]bool{}
 	)
 
-	// TODO(adam): implement planner logic that will push down a matching call to distinct() into this call, setting t.distinct to true
-	if t.distinct {
-		switch keyColType {
-		case flux.TBool:
-			boolDistinct = make(map[bool]bool)
-		case flux.TInt:
-			intDistinct = make(map[int64]bool)
-		case flux.TUInt:
-			uintDistinct = make(map[uint64]bool)
-		case flux.TFloat:
-			floatDistinct = make(map[float64]bool)
-		case flux.TString:
-			stringDistinct = make(map[string]bool)
-		case flux.TTime:
-			timeDistinct = make(map[execute.Time]bool)
-		}
-	}
-
 	return tbl.Do(func(cr flux.ColReader) error {
 		l := cr.Len()
 		for i := 0; i < l; i++ {
 			// Check distinct
-			for j, rowIdx := range keyColIndices {
-				if rowIdx < 0 {
-					continue
-				}
+			for _, c := range keyColumns {
 				switch keyColType {
 				case flux.TBool:
-					vs := cr.Bools(rowIdx)
+					vs := cr.Bools(c.idx)
```

Comment:
I started reviewing this patch, got to this point, and thought "Am I reviewing generated code?" Kudos to you for manually handling this giant type switch.

Reply:
No problem.